flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/runtime/stream/sql/UnnestITCase.scala

```diff
@@ -312,71 +306,509 @@ class UnnestITCase(mode: StateBackendMode) extends StreamingWithStateTestBase(mo
+      isRetract = false,
+      parallelism = -1,
+      fieldNames = 'a,
+      'b,
+      'c
+    )
+  }
+
+  @TestTemplate
+  def testUnnestArrayOfArrayWithOrdinality(): Unit = {
+    val data = List(
+      (1, Array(Array(1, 2), Array(3))),
+      (2, Array(Array(4, 5), Array(6, 7))),
+      (3, Array(Array(8)))
+    )
+    assertUnnest(
+      testData = data,
+      typeInfo = createTypeInformation[(Int, Array[Array[Int]])],
+      sqlQuery = """
+                   |SELECT id, array_val, array_pos, `element`, element_pos
+                   |FROM T
+                   |CROSS JOIN UNNEST(nested_array) WITH ORDINALITY AS A(array_val, array_pos)
+                   |CROSS JOIN UNNEST(array_val) WITH ORDINALITY AS B(`element`, element_pos)
+                   |""".stripMargin,
+      expectedResults = List(
+        "1,[1, 2],1,1,1",
+        "1,[1, 2],1,2,2",
+        "1,[3],2,3,1",
+        "2,[4, 5],1,4,1",
+        "2,[4, 5],1,5,2",
+        "2,[6, 7],2,6,1",
+        "2,[6, 7],2,7,2",
+        "3,[8],1,8,1"),
+      isRetract = false,
+      parallelism = -1,
+      fieldNames = 'id,
+      'nested_array
+    )
+  }
+
+  @TestTemplate
+  def testUnnestMultisetWithOrdinality(): Unit = {
+    val data = List(
+      (1, 1, "Hi"),
+      (1, 2, "Hello"),
+      (2, 2, "World"),
+      (3, 3, "Hello world")
+    )
+    assertUnnest(
+      testData = data,
+      typeInfo = createTypeInformation[(Int, Int, String)],
+      sqlQuery = """
+                   |WITH T1 AS (SELECT a, COLLECT(c) as words FROM T GROUP BY a)
+                   |SELECT a, word, pos
+                   |FROM T1 CROSS JOIN UNNEST(words) WITH ORDINALITY AS A(word, pos)
+                   |""".stripMargin,
+      expectedResults = List(
+        "1,Hi,1",
+        "1,Hello,2",
+        "2,World,1",
+        "3,Hello world,1"
+      ),
+      isRetract = true,
+      parallelism = -1,
+      fieldNames = 'a,
+      'b,
+      'c
+    )
+  }
+
+  @TestTemplate
+  def testUnnestMapWithOrdinality(): Unit = {
+    val data = List(
+      Row.of(
+        Int.box(1), {
+          val map = new java.util.HashMap[String, String]()
+          map.put("a", "10")
+          map.put("b", "11")
+          map
+        }),
+      Row.of(
+        Int.box(2), {
+          val map = new java.util.HashMap[String, String]()
+          map.put("c", "20")
+          map.put("d", "21")
+          map
+        })
+    )
+
+    implicit val typeInfo = Types.ROW(
+      Array("id", "map_data"),
+      Array[TypeInformation[_]](Types.INT, Types.MAP(Types.STRING, Types.STRING))
+    )
+
+    assertUnnest(
+      testData = data,
+      typeInfo = typeInfo,
+      sqlQuery = """
+                   |SELECT id, k, v, pos
+                   |FROM T CROSS JOIN UNNEST(map_data) WITH ORDINALITY AS f(k, v, pos)
+                   |""".stripMargin,
+      expectedResults = List("1,a,10,1", "1,b,11,2", "2,c,20,2", "2,d,21,1"),
+      isRetract = false,
+      parallelism = 1,
```

Comment:
Is this parallelism important for the behavior of the test? if not, we can run all tests with the same value and remove this parameter from `assertUnnest()`